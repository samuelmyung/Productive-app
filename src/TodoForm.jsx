import React, { useState } from "react";

const DEFAULT_TODO_DATA = {
  title: '',
  description: '',
  priority: 3
};

/** Form for adding.
 *
 * Props:
 * - initialFormData
 * - handleSave: function to call in parent.
 *
 * { TodoApp, EditableTodo } -> TodoForm
 */

function TodoForm({ handleSave, initialFormData = DEFAULT_TODO_DATA }) {
  const [formData, setFormData] = useState(initialFormData);
  const [titleIsValid, setTitleIsValid] = useState(true);

  /** Update form input. */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(data => ({
      ...data,
      [name]: value
    }));
  }

  /** Call parent function and clear form. */
  function handleSubmit(evt) {
    evt.preventDefault();

    const titleValidity = formData.title.length > 0;
    setTitleIsValid(titleValidity);
    if (!titleValidity) return;

    handleSave(formData);
    setFormData(initialFormData);
  }


  return (
    <form className="NewTodoForm" onSubmit={handleSubmit}>

      <div className="mb-3">
        <input
          id="newTodo-title"
          name="title"
          className="form-control"
          placeholder="Title"
          onChange={handleChange}
          value={formData.title}
          aria-label="Title"
        />
      </div>

      {!titleIsValid && <p style={{color: "red"}}>Title required!</p>}

      <div className="mb-3">
        <textarea
          id="newTodo-description"
          name="description"
          className="form-control"
          placeholder="Description"
          onChange={handleChange}
          value={formData.description}
          aria-label="Description"
        />
      </div>

      <div className="mb-3 d-flex justify-content-between">
        <div className="w-75 d-flex justify-content-between">
          <label htmlFor="newTodo-priority"
            className="d-inline-flex">Priority:&nbsp;&nbsp;
          </label>
          <select id="newTodo-priority"
            name="priority"
            value={formData.priority}
            onChange={handleChange}
            className="form-control form-control-sm d-inline-flex"
          >
            <option value={1}>Ultra-Über</option>
            <option value={2}>Über</option>
            <option value={3}>Meh</option>
          </select>
        </div>
        <button className="btn-primary rig btn btn-sm NewTodoForm-addBtn">
          Gø!
        </button>
      </div>

    </form>
  );
}

export default TodoForm;
