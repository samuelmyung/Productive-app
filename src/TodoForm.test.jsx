import { describe, it, expect, vi } from "vitest";

import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import TodoForm from "./TodoForm";

it("renders without crashing", function () {
    render(<TodoForm />);
});

it("matches initial form snapshot", function () {
    const { asFragment } = render(<TodoForm />);
    expect(asFragment()).toMatchSnapshot();
});

it("matches data-entered snapshot", function () {
    const { container } = render(<TodoForm />);

    const titleInput = container.querySelector('[name=title]');
    const descriptionInput = container.querySelector('[name=description]');
    const priorityInput = container.querySelector('[name=priority]');

    fireEvent.change(titleInput, { target: { value: 'test title' }});
    fireEvent.change(descriptionInput, { target: { value: 'test description' }});
    fireEvent.change(priorityInput, { target: { value: 1 }});

    expect(container).toMatchSnapshot();
});

it("submitting form ", function () {
    const mockCreateTodo = vi.fn();
    const { container } = render(<TodoForm handleSave={mockCreateTodo} />);

    const titleInput = container.querySelector('[name=title]');
    const descriptionInput = container.querySelector('[name=description]');
    const priorityInput = container.querySelector('[name=priority]');

    fireEvent.input(titleInput, { target: { value: 'test title' }});
    fireEvent.input(descriptionInput, { target: { value: 'test description' }});
    fireEvent.input(priorityInput, { target: { value: 1 }});

    expect(mockCreateTodo).toHaveBeenCalledTimes(0);

    fireEvent.click(container.querySelector('.NewTodoForm-addBtn'));
    expect(mockCreateTodo).toHaveBeenCalledTimes(1);

    expect(container.querySelectorAll("input[value='']")).toHaveLength(1);
    expect(container.querySelectorAll("textarea")[0].value).toEqual('');

    const options = container.querySelectorAll('option');
    // expect(options[0].selected).toBeTruthy();
    // expect(container.querySelectorAll("select")[0].value).toEqual('1');
});