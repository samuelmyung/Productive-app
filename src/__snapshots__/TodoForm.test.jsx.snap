// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches data-entered snapshot 1`] = `
<div>
  <form
    class="NewTodoForm"
  >
    <div
      class="mb-3"
    >
      <input
        aria-label="Title"
        class="form-control"
        id="newTodo-title"
        name="title"
        placeholder="Title"
        value="test title"
      />
    </div>
    <div
      class="mb-3"
    >
      <textarea
        aria-label="Description"
        class="form-control"
        id="newTodo-description"
        name="description"
        placeholder="Description"
      >
        test description
      </textarea>
    </div>
    <div
      class="mb-3 d-flex justify-content-between"
    >
      <div
        class="w-75 d-flex justify-content-between"
      >
        <label
          class="d-inline-flex"
          for="newTodo-priority"
        >
          Priority:  
        </label>
        <select
          class="form-control form-control-sm d-inline-flex"
          id="newTodo-priority"
          name="priority"
        >
          <option
            value="1"
          >
            Ultra-Über
          </option>
          <option
            value="2"
          >
            Über
          </option>
          <option
            value="3"
          >
            Meh
          </option>
        </select>
      </div>
      <button
        class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
      >
        Gø!
      </button>
    </div>
  </form>
</div>
`;

exports[`matches initial form snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewTodoForm"
  >
    <div
      class="mb-3"
    >
      <input
        aria-label="Title"
        class="form-control"
        id="newTodo-title"
        name="title"
        placeholder="Title"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <textarea
        aria-label="Description"
        class="form-control"
        id="newTodo-description"
        name="description"
        placeholder="Description"
      />
    </div>
    <div
      class="mb-3 d-flex justify-content-between"
    >
      <div
        class="w-75 d-flex justify-content-between"
      >
        <label
          class="d-inline-flex"
          for="newTodo-priority"
        >
          Priority:  
        </label>
        <select
          class="form-control form-control-sm d-inline-flex"
          id="newTodo-priority"
          name="priority"
        >
          <option
            value="1"
          >
            Ultra-Über
          </option>
          <option
            value="2"
          >
            Über
          </option>
          <option
            value="3"
          >
            Meh
          </option>
        </select>
      </div>
      <button
        class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
      >
        Gø!
      </button>
    </div>
  </form>
</DocumentFragment>
`;
